### PR TITLE
Implement the itemDependencyApply service

### LINKED ISSUE #695: item-dependencies: apply to existing
## Overview 
(re)Apply a specific existing item-dependencies rule on existing results

On `POST /items/{dependent_item_id}/prerequisites/{prerequisite_item_id}/grant_content_view`

## Input
* dependent_item_id, prerequisite_item_id

## Validations/Restrictions
- The dependent_item_id, prerequisite_item_id must exist, with grant_content_view=1 
- current-user must have:
  - can_edit=all on dependent_item_id
  - can_grant_view >= content on the dependent_item_id

## Actions
Apply the rule, i.e. grant the content access, for all existing participants which meet the condition defined by this dependency.
Do not delete entries for those who would not meet the condition anymore.

## Output
Generic output

### PR DESCRIPTION
**Pull request by zenovich**
Fixes #695

### CODE REVIEW

File: app/database/result_store_propagate.go [hunk 1]
```diff
@@ -302,9 +301,8 @@ func (s *ResultStore) Propagate() (err error) {
 				can_enter_from = IF(
 					VALUES(can_enter_from) <> '9999-12-31 23:59:59' AND can_enter_from > VALUES(can_enter_from),
 					VALUES(can_enter_from), can_enter_from),
-				can_enter_until = IF(
-					VALUES(can_enter_from) <> '9999-12-31 23:59:59' AND can_enter_until <> '9999-12-31 23:59:59',
-					'9999-12-31 23:59:59', can_enter_until)`, canViewContentIndex, canViewContentIndex)
+				can_enter_until = IF(VALUES(can_enter_from) <> '9999-12-31 23:59:59', '9999-12-31 23:59:59', can_enter_until)`,
```
Comment (smadbe) [inline app/database/result_store_propagate.go:304]: But so, if the new value of `from`is +inf, you set `until` to its existing value? It looks like the case where you want to put `until` to `+inf` anyway. (`until` < `from` makes no sense).

The old rule meant: we set `until` to `+inf` EXCEPT if it previous value was not `+inf`and the new `from` is `+inf`... in which case we leave it to its previous value. (so `until` < `from` )

But as this `until` < `from`  does not make sense and "item_unlocking" rules are only set by code that we control, I think, we can just ignore `can_enter_until` and leave it to its db default value "9999-12-31 23:59:59".

**Review by smadbe — CHANGES_REQUESTED** [review]

File: app/database/result_store_propagate.go (hunk 1 above)
Comment (zenovich) [reply app/database/result_store_propagate.go:304]: In this query we never want to set `can_enter_until` to +inf explicitly.

Actually, the old rule meant we set until to +inf EXCEPT if it previous value was not +inf OR the new from is +inf... in which case we leave it to its previous value.

The only difference between the old rule and the new rule is that the old rule explicitly kept the old value of `can_enter_until` if it was +inf. The new rule overwrites +inf by +inf. So there is no difference between the rules.

Comment (smadbe) [reply app/database/result_store_propagate.go:304]: I just mean, that the new like the new rule does not really sense IMO. So we may not set `can_enter_until` at all.

Comment (zenovich) [reply app/database/result_store_propagate.go:304]: But if the can_enter_until is in the past, the item will not be unlocked (no matter how we set can_enter_from), so we should update can_enter_until here.

Comment (smadbe) [reply app/database/result_store_propagate.go:304]: Why would it be in the past for a "item_unlocking" row? But so, put it to "9999-12-31 23:59:59" without conditions.

Comment (zenovich) [reply app/database/result_store_propagate.go]: fixed

**Review by smadbe — APPROVED** [review]

**Pull request merged**